File: Palindrome Partitioning.cpp
class Solution {
public:
    // concept used : substring , palindrome , recursion
    vector<vector<string>> partition(string s) {
        vector<vector<string>> res;
        vector<string> ds;
        find(s,res,ds,0);
        return res;
    }
    void find(string s,vector<vector<string>>&res,vector<string>&ds,int ind){
        if(ind==s.size()) 
        {
            res.push_back(ds);
            return;
        }
        for(int i=ind;i<s.size();i++){
          if(ispal(s,ind,i)){
              ds.push_back(s.substr(ind,i-ind+1));
              find(s,res,ds,i+1);
              ds.pop_back();
          }
        }
    }
    bool ispal(string s,int st,int en){
      while(st<=en){
          if(s[st++]!=s[en--]){
              return false;
          }
      }
      return true;
    }
};

File: Palindrome Partitioning.java
class Solution {
    public boolean isPalindrome(String s){
        int i = 0, j = s.length()-1;

        while(i < j){
            if(s.charAt(i)!=s.charAt(j)){
                return false;
            }
            i++;
            j--;
        }

        return true;
    }

    public void helper(int idx, String s, String tempStr, List<List<String>> res, List<String> ls ){
        if(idx == s.length()) {
            res.add(new ArrayList<>(ls));
            return;
        }

        tempStr = "";
        
        for(int i = idx; i<s.length(); i++){
            tempStr += ""+s.charAt(i);
            if(isPalindrome(tempStr)){
                ls.add(tempStr);
                helper(i+1, s, "", res, ls);
                ls.remove(ls.size()-1);
            }
        }
    }

    public List<List<String>> partition(String s) {
        List<List<String>> res = new ArrayList<>();
        helper(0, s , "", res, new ArrayList<>());
        return res;
    }
}
